smartsim/_core/mli/infrastructure/storage/backbonefeaturestore.py

```diff
@@ -0,0 +1,381 @@
+        logger.debug(f"Broadcasting {num_pending} events to {num_consumers} consumers")
+
+    def _prune_unused_consumers(self) -> None:
+        """Performs maintenance on the channel cache by pruning any channel
+        that has been removed from the consumers list"""
+        active_consumers = set(self._descriptors)
+        current_channels = set(self._channel_cache.keys())
+
+        # find any cached channels that are now unused
+        inactive_channels = current_channels.difference(active_consumers)
+        new_channels = active_consumers.difference(current_channels)
+
+        for descriptor in inactive_channels:
+            self._channel_cache.pop(descriptor)
+
+        logger.debug(
+            f"Pruning {len(inactive_channels)} stale consumer channels"
+            f" and found {len(new_channels)} new channels"
+        )
+
+    def _get_comm_channel(self, descriptor: str) -> CommChannelBase:
+        """Helper method to build and cache a comm channel
+
+        :param descriptor: the descriptor to pass to the channel factory
+        :returns: the instantiated channel
+        :raises SmartSimError: if the channel fails to build"""
+        comm_channel = self._channel_cache[descriptor]
+        if comm_channel is not None:
+            return comm_channel
+
+        if self._channel_factory is None:
+            raise SmartSimError("No channel factory provided for consumers")
+
+        try:
+            channel = self._channel_factory(descriptor)
+            self._channel_cache[descriptor] = channel
+            return channel
+        except Exception as ex:
+            msg = f"Unable to construct channel with descriptor: {descriptor}"
+            logger.error(msg, exc_info=True)
+            raise SmartSimError(msg) from ex
+
+    def _broadcast(self) -> int:
+        """Broadcasts all buffered events to registered event consumers.
+
+        :return: the number of events broadcasted to consumers
+        :raises ValueError: if event serialization fails
+        :raises KeyError: if channel fails to attach using registered descriptors
+        :raises SmartSimError: if broadcasting fails"""
+
+        # allow descriptors to be empty since events are buffered
+        self._descriptors = set(x for x in self._backbone.notification_channels if x)
+        if not self._descriptors:
+            logger.warning("No event consumers are registered")
+            return 0
+
+        self._prune_unused_consumers()
+        self._log_broadcast_start()
+
+        num_sent: int = 0
+        next_event: t.Optional[bytes] = self._event_buffer.popleft()
+
+        # send each event to every consumer
+        while next_event is not None:
+            for descriptor in map(str, self._descriptors):
+                comm_channel = self._get_comm_channel(descriptor)
+
+                try:
+                    comm_channel.send(next_event)
+                    num_sent += 1
+                except Exception as ex:
```

Comment:
If one channel broadcast fails, then an error is raised.  This means that other channels that may be OK will never get the message because it was already popped and not placed back on the deque.  I wonder if we should stored failed sends and still go through other channels?  We would still need to think about if the failed broadcasts should be retired or just thrown away...

Reply:
per offline discussion, handling in follow-up PR